src/internal/core.rs

```diff
@@ -30,6 +30,8 @@ pub struct State<P: Package, VS: VersionSet, Priority: Ord + Clone> {
     /// and will stay that way until the next conflict and backtrack is operated.
     contradicted_incompatibilities: rustc_hash::FxHashSet<IncompId<P, VS>>,
 
+    dependencies: Map<(P, P), SmallVec<IncompId<P, VS>>>,
```

Comment:
I guess you already tested it but I'm gonna ask just in case. Considering we only merge dependants where the dependency is exactly the same, would it make sense to store the dependency in the key? It does increase memory so it's probably worse, but it would prevent from looping through all incompats for the pair.

Reply:
I suppose these package pairs are generally small anyway since you opted for `SmallVec` so yeah maybe it's a bad trade off. But still curious to hear from if you tried it.

Reply:
> Considering we only merge dependants where the dependency is exactly the same, would it make sense to store the dependency in the key? It does increase memory so it's probably worse, but it would prevent from looping through all incompats for the packages pair.

I would absolutely love to do this. Unfortunately `VS` is neither `Hash` nor `Ord`, so I don't know of an efficient data structure for using them as a key.

Reply:
Ah right, `Ord` would not make sense for sets. `Hash` could make sense. Maybe we can make a note of this, to explore in the future if it turns out some real-world use case end up being in the worst case scenario for this (many different ranges for a package pair, causing O(n²) complexity).

Reply:
Realistically, this can only happen for a package pair  where there are many versions of each, their versions are bumped in sync. Something like two packages from the same workspace maybe if versions are strictly specified. Or for things like features (in our "pubgrub limitations" guide section) where we pair exactly a feature version to the bare crate version. Hum, maybe not so unlikely ...

Reply:
Yes. "Features" is why none of the cargo benchmarks hit this code path.

I'm not too afraid of the O(n²) overhead here. Every time we merge dependencies we follow it up by doing `unit_propagation`, which does a linear scan over an even larger collection of incompatibilities. That is to say it has the same O(n²) shaped, with a larger n. If this becomes a hotspot in a resolution that is slow we can investigate a fixed then.